### PR TITLE
📝 For review [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by Vectorized**
## Description

Describe the changes made in your pull request here.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [ ] Ran `forge fmt`?
- [ ] Ran `forge snapshot`?
- [ ] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

File: src/PreApproveChecker.sol [hunk 1]
```diff
@@ -1,57 +1,57 @@
...
-        }
-    }
-}
+// SPDX-License-Identifier: MIT 
+pragma solidity ^0.8.4; 
+ 
+/** 
+ * @title PreApproveChecker 
+ * @notice A library for checking whether an operator is pre-approved 
+ *         by a lister subcribed to by a collector.
```
Comment (gigamesh) [inline src/PreApproveChecker.sol:7]:
```suggestion
 *         by a lister subscribed to by a collector. 
```

File: src/PreApproveRegistry.sol [hunk 2]
```diff
@@ -1,304 +1,304 @@
...
+    // ============================================================= 
+ 
+    constructor() payable {} 
+ 
+    // ============================================================= 
+    //               PUBLIC / EXTERNAL WRITE FUNCTIONS 
+    // ============================================================= 
+ 
+    /** 
+     * @dev Subscribes the caller (collector) from `lister`.
```
Comment (gigamesh) [inline src/PreApproveRegistry.sol:103]:
```suggestion
     * @dev Subscribes the caller (collector) to `lister`. 
```

File: src/utils/EnumerableAddressSetMap.sol [hunk 3]
```diff
@@ -1,186 +1,186 @@
...
+     * @dev Returns all the values of the enumerable address set 
+     *      at `key` in `map`. 
+     * @param map The mapping of enumerable address sets. 
+     * @param key The key into the mapping. 
+     * @return The values. 
+     */ 
+    function values(Map storage map, address key) internal view returns (address[] memory) { 
+        return map._values[key]; 
+    } 
+}
```
Comment (gigamesh) [inline src/utils/EnumerableAddressSetMap.sol:186]: I feel like it would be safer to make this lib a separate PR with a diff against the OZ contract for easier reviewing.

However, if everyone else is comfortable with including here, ignore my comment.

File: test/utils/TestPlus.sol [hunk 4]
```diff
@@ -1,162 +1,162 @@
...
+                if iszero(r) { 
+                    result := min 
+                    break 
+                } 
+                result := add(sub(max, r), 1) 
+                break 
+            } 
+        } 
+    } 
+}
```
Comment (gigamesh) [inline test/utils/TestPlus.sol:162]: Where do these functions get used?

**Review by gigamesh — CHANGES_REQUESTED** [review]

File: test/utils/TestPlus.sol (hunk 4 above)
Comment (Vectorized) [reply test/utils/TestPlus.sol:162]: ^ Oh... this the test base class copypasted from Solady.

Specifically, I need the `_random()` function.

File: src/utils/EnumerableAddressSetMap.sol (hunk 3 above)
Comment (Vectorized) [reply src/utils/EnumerableAddressSetMap.sol:186]: This lib is specifically only for this registry, as it abuses `returndatasize()` instead of `0`.

This trick can only be used if the entire contract does not make any external calls or staticcalls.

File: src/PreApproveLister.sol [hunk 5]
```diff
@@ -1,82 +1,82 @@
...
+     * @param operator The account that can manage NFTs on behalf of 
+     *                 collectors subscribed to the caller. 
+     */ 
+    function addOperator(address operator) external payable onlyOwner { 
+        /// @solidity memory-safe-assembly 
+        assembly { 
+            // Store the function selector. 
+            mstore(returndatasize(), calldataload(returndatasize())) 
+            // Store the operator. 
+            mstore(0x04, operator)
```
Comment (moodlezoup) [inline src/PreApproveLister.sol:42]: this is cheaper than `calldatacopy`?

Comment (Vectorized) [reply src/PreApproveLister.sol:42]: Good idea.
